applications/HDF5Application/custom_io/hdf5_file.h

```diff
@@ -85,216 +99,409 @@ class KRATOS_API(HDF5_APPLICATION) File
+    template<class TDataType>
+    void ReadDataSet(
+        const std::string& rPath,
+        TDataType& rData,
+        const unsigned StartIndex,
+        const unsigned BlockSize);
 
-    virtual void ReadDataSet(const std::string& rPath,
-                             Vector<double>& rData,
-                             unsigned StartIndex,
-                             unsigned BlockSize);
-
-    virtual void ReadDataSet(const std::string& rPath,
-                             Vector<array_1d<double, 3>>& rData,
-                             unsigned StartIndex,
-                             unsigned BlockSize);
-
-    virtual void ReadDataSet(const std::string& rPath,
-                             Matrix<int>& rData,
-                             unsigned StartIndex,
-                             unsigned BlockSize);
-
-    virtual void ReadDataSet(const std::string& rPath,
-                             Matrix<double>& rData,
-                             unsigned StartIndex,
-                             unsigned BlockSize);
-
-    // Independently read data set from the HDF5 file.
     /**
-     *  Performs independent read in MPI. Throws if out of range.
+     * @brief Independently read a data set from the HDF5 file.
+     *
+     * Performs independent read in MPI. Throws if out of range.
+     *
+     * @tparam TDataType        Datatype of the read data.
+     * @param rPath             Path of the dataset.
+     * @param rData             Data to be written to.
+     * @param StartIndex        Starting offset of data for this rank.
+     * @param BlockSize         Number of data points for this rank.
      */
-    virtual void ReadDataSetIndependent(const std::string& rPath,
-                                       Vector<int>& rData,
-                                       unsigned StartIndex,
-                                       unsigned BlockSize);
-
-    virtual void ReadDataSetIndependent(const std::string& rPath,
-                                       Vector<double>& rData,
-                                       unsigned StartIndex,
-                                       unsigned BlockSize);
-
-    virtual void ReadDataSetIndependent(const std::string& rPath,
-                                       Vector<array_1d<double, 3>>& rData,
-                                       unsigned StartIndex,
-                                       unsigned BlockSize);
-
-    virtual void ReadDataSetIndependent(const std::string& rPath,
-                                       Matrix<int>& rData,
-                                       unsigned StartIndex,
-                                       unsigned BlockSize);
-
-    virtual void ReadDataSetIndependent(const std::string& rPath,
-                                       Matrix<double>& rData,
-                                       unsigned StartIndex,
-                                       unsigned BlockSize);
+    template<class TDataType>
+    void ReadDataSetIndependent(
```

Comment:
Same suggestion to pass `Independent` or `Collective` as a function argument or template parameter.

Reply:
It is already like that, this is kepy for backward compatibiltiy.